src/values/AbstractValue.js

```diff
@@ -321,31 +322,50 @@ export default class AbstractValue extends Value {
-      let refinedRight = right instanceof AbstractValue ? right.refineWithPathCondition() : right;
-      if (left === refinedLeft && right === refinedRight) return this;
-      return AbstractValue.createFromLogicalOp(this.$Realm, op, refinedLeft, refinedRight, this.expressionLocation);
+    function pathImplies(condition: AbstractValue): boolean {
+      let path = condition.$Realm.pathConditions;
+      for (let i = path.length - 1; i >= 0; i--) {
+        let pathCondition = path[i];
+        if (pathCondition.implies(condition)) return true;
+      }
+      return false;
     }
-    if (op !== "conditional") return this;
-    let [condition, trueVal, falseVal] = this.args;
-    invariant(condition instanceof AbstractValue);
-    invariant(trueVal !== undefined);
-    invariant(falseVal !== undefined);
-    let inverseCondition = AbstractValue.createFromUnaryOp(this.$Realm, "!", condition);
-    let result = this;
-    for (let pathCondition of this.$Realm.pathConditions) {
-      if (pathCondition.implies(condition)) {
-        result = trueVal;
-        break;
+
+    let realm = this.$Realm;
+    let op = this.kind;
+    let result = (() => {
+      if (op === "&&" || op === "||") {
+        let [left, right] = this.args;
+        let refinedLeft = left instanceof AbstractValue ? left.refineWithPathCondition() : left;
+        let refinedRight = right instanceof AbstractValue ? right.refineWithPathCondition() : right;
+        if (left === refinedLeft && right === refinedRight) return this;
+        return AbstractValue.createFromLogicalOp(realm, op, refinedLeft, refinedRight, this.expressionLocation);
       }
-      if (pathCondition.implies(inverseCondition)) {
-        result = falseVal;
-        break;
+      if (op === "!") {
+        let arg = this.args[0];
+        let refinedArg = arg instanceof AbstractValue ? arg.refineWithPathCondition() : arg;
+        if (arg === refinedArg) return this;
+        if (!refinedArg.mightNotBeTrue()) return realm.intrinsics.false;
+        if (!refinedArg.mightNotBeFalse()) return realm.intrinsics.true;
+        invariant(refinedArg instanceof AbstractValue); // concrete values always make up their mind above
+        return AbstractValue.createFromUnaryOp(realm, op, refinedArg);
       }
-    }
+      if (op !== "conditional") return this;
+      let [condition, trueVal, falseVal] = this.args;
+      invariant(trueVal !== undefined);
+      invariant(falseVal !== undefined);
+      invariant(condition instanceof AbstractValue);
+      let inverseCondition = AbstractValue.createFromUnaryOp(this.$Realm, "!", condition);
+      if (pathImplies(condition)) return trueVal;
+      if (pathImplies(inverseCondition)) return falseVal;
+      if (trueVal instanceof NullValue || trueVal instanceof UndefinedValue) {
+        if (pathImplies(AbstractValue.createFromBinaryOp(realm, "!==", this, trueVal))) return falseVal;
```

Comment:
Why is this check nested in a `trueVal instanceof NullValue || trueVal instanceof UndefinedValue` branch? Doesn't it always make sense? Or is it to avoid executing some kind that has side effects?

Reply:
What is happening here is that we may have absval = "some not so nice condition" ? trueVal : falseVal. Absval can be either trueVal or falseVal, so if it is not trueVal, it must be falseVal. We have already failed to work out that the path implies "some not so nice condition", so if we can instead work out the the path implies "absVal !== trueVal", we can stay in the refinement business. It took me a while to remember this, so I'll add a comment to help future readers.